File: lib/services/openai_service.dart
/// Direct OpenAI API service for chat
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'settings_service.dart';

class OpenAIService {
  final String apiKey;
  final String model;
  
  OpenAIService({
    required this.apiKey,
    this.model = 'gpt-5-mini',
  });

  /// Chat with OpenAI using conversation context
  Future<String> chat({
    required String userMessage,
    String? conversationContext,
  }) async {
    try {
      final messages = <Map<String, String>>[];
      
      // System message with context
      String systemPrompt = 'You are a helpful AI assistant. You have access to the user\'s conversation history and memories.';
      
      if (conversationContext != null && conversationContext.isNotEmpty) {
        systemPrompt += '''

Here is the user's recent conversation history for context:

$conversationContext

Use this context to provide personalized and relevant responses. Reference specific conversations when appropriate.''';
      }
      
      messages.add({
        'role': 'system',
        'content': systemPrompt,
      });
      
      messages.add({
        'role': 'user',
        'content': userMessage,
      });

      final response = await http.post(
        Uri.parse('https://api.openai.com/v1/chat/completions'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $apiKey',
        },
        body: jsonEncode({
          'model': model,
          'messages': messages,
          'max_tokens': 1000,
        }),
      );

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body);
        final content = json['choices']?[0]?['message']?['content'];
        
        // Track token usage
        final usage = json['usage'];
        if (usage != null) {
          SettingsService.addOpenAIUsage(
            usage['prompt_tokens'] ?? 0,
            usage['completion_tokens'] ?? 0,
          );
        }
        
        return content ?? 'No response generated';
      } else {
        debugPrint('OpenAI API error: ${response.statusCode} ${response.body}');
        throw Exception('OpenAI API error: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('OpenAI chat error: $e');
      rethrow;
    }
  }

  /// Generate a title, summary, extract memories and tasks from a conversation
  Future<Map<String, dynamic>> summarizeConversation(String transcript, {DateTime? currentTime}) async {
    final now = currentTime ?? DateTime.now();
    final timeContext = 'Current date/time: ${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')} ${now.hour.toString().padLeft(2, '0')}:${now.minute.toString().padLeft(2, '0')}';
    
    try {
      final response = await http.post(
        Uri.parse('https://api.openai.com/v1/chat/completions'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $apiKey',
        },
        body: jsonEncode({
          'model': model,
          'messages': [
            {
              'role': 'system',
              'content': '''You analyze conversations and extract key information.
$timeContext

Respond with JSON only:
{
  "title": "short descriptive title",
  "summary": "brief 1-2 sentence summary",
  "memories": ["important fact 1", "important fact 2"],
  "tasks": [
    {"title": "task description", "due_date": "2024-12-12T18:00:00"}
  ]
}

For memories, extract ONLY important facts worth remembering long-term, such as:
- Names (e.g., "User's name is Karsten")
- Preferences (e.g., "User prefers tea over coffee")
- Personal details (e.g., "User works as a software engineer")

For tasks, extract actionable items mentioned:
- Things the user needs to do (e.g., "I have to write my essay tonight" → task with due_date tonight around 6pm)
- Appointments or deadlines mentioned
- Use ISO 8601 format for due_date (or null if no time mentioned)
- Infer reasonable times: "tonight" = 6pm today, "tomorrow morning" = 9am tomorrow

If there are no notable facts/tasks, return empty arrays.
Keep each item as a short, clear statement.'''
            },
            {
              'role': 'user',
              'content': 'Analyze this conversation:\n\n$transcript'
            }
          ],
          'max_tokens': 700,
          'response_format': {'type': 'json_object'},
        }),
      );

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body);
        final content = json['choices']?[0]?['message']?['content'];
        
        // Track token usage
        final usage = json['usage'];
        if (usage != null) {
          SettingsService.addOpenAIUsage(
            usage['prompt_tokens'] ?? 0,
            usage['completion_tokens'] ?? 0,
          );
        }
        
        if (content != null) {
          final parsed = jsonDecode(content);
          return {
            'title': parsed['title'] ?? 'Untitled Conversation',
            'summary': parsed['summary'] ?? '',
            'memories': (parsed['memories'] as List?)?.cast<String>() ?? [],
            'tasks': parsed['tasks'] ?? [],
          };
        }
      }
      return {'title': 'Untitled Conversation', 'summary': '', 'memories': <String>[], 'tasks': []};
    } catch (e) {
      debugPrint('OpenAI summarize error: $e');
      return {'title': 'Untitled Conversation', 'summary': '', 'memories': <String>[], 'tasks': []};
    }
  }
}

File: OmiLocal/Sources/Services/OpenAIService.swift
import Foundation

class OpenAIService {
    let apiKey: String
    let model: String
    
    init(apiKey: String, model: String = "gpt-4.1-mini") {
        self.apiKey = apiKey
        self.model = model
    }
    
    func chat(userMessage: String, conversationContext: String?) async throws -> String {
        var messages: [[String: String]] = []
        
        var systemPrompt = "You are a helpful AI assistant. You have access to the user's conversation history and memories."
        
        if let context = conversationContext, !context.isEmpty {
            systemPrompt += """
            
            Here is the user's recent conversation history for context:
            
            \(context)
            
            Use this context to provide personalized and relevant responses.
            """
        }
        
        messages.append(["role": "system", "content": systemPrompt])
        messages.append(["role": "user", "content": userMessage])
        
        guard let url = URL(string: "https://api.openai.com/v1/chat/completions") else {
            throw NSError(domain: "OpenAI", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        let body: [String: Any] = [
            "model": model,
            "messages": messages,
            "max_tokens": 1000
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw NSError(domain: "OpenAI", code: 2, userInfo: [NSLocalizedDescriptionKey: "API request failed"])
        }
        
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = json["choices"] as? [[String: Any]],
              let firstChoice = choices.first,
              let message = firstChoice["message"] as? [String: Any],
              let content = message["content"] as? String else {
            throw NSError(domain: "OpenAI", code: 3, userInfo: [NSLocalizedDescriptionKey: "Failed to parse response"])
        }
        
        if let usage = json["usage"] as? [String: Any] {
            let inputTokens = usage["prompt_tokens"] as? Int ?? 0
            let outputTokens = usage["completion_tokens"] as? Int ?? 0
            SettingsService.shared.addOpenAIUsage(input: inputTokens, output: outputTokens)
        }
        
        return content
    }
    
    func summarizeConversation(_ transcript: String, currentTime: Date = Date()) async throws -> SummaryResult {
        let dateFormatter = ISO8601DateFormatter()
        let timeContext = "Current date/time: \(dateFormatter.string(from: currentTime))"
        
        guard let url = URL(string: "https://api.openai.com/v1/chat/completions") else {
            throw NSError(domain: "OpenAI", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        let systemPrompt = """
        You analyze conversations and extract key information.
        \(timeContext)
        
        Respond with JSON only:
        {
          "title": "short descriptive title",
          "summary": "brief 1-2 sentence summary",
          "memories": ["important fact 1", "important fact 2"],
          "tasks": [
            {"title": "task description", "due_date": "2024-12-12T18:00:00"}
          ]
        }
        
        For memories, extract ONLY important facts worth remembering long-term, such as:
        - Names (e.g., "User's name is Karsten")
        - Preferences (e.g., "User prefers tea over coffee")
        - Personal details (e.g., "User works as a software engineer")
        
        For tasks, extract actionable items mentioned.
        
        If there are no notable facts/tasks, return empty arrays.
        """
        
        let body: [String: Any] = [
            "model": model,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user", "content": "Analyze this conversation:\n\n\(transcript)"]
            ],
            "max_tokens": 700,
            "response_format": ["type": "json_object"]
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw NSError(domain: "OpenAI", code: 2, userInfo: [NSLocalizedDescriptionKey: "API request failed"])
        }
        
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = json["choices"] as? [[String: Any]],
              let firstChoice = choices.first,
              let message = firstChoice["message"] as? [String: Any],
              let content = message["content"] as? String else {
            return SummaryResult(title: "Untitled Conversation", summary: "", memories: [], tasks: [])
        }
        
        guard let parsed = try? JSONSerialization.jsonObject(with: Data(content.utf8)) as? [String: Any] else {
            return SummaryResult(title: "Untitled Conversation", summary: "", memories: [], tasks: [])
        }
        
        let title = parsed["title"] as? String ?? "Untitled Conversation"
        let summary = parsed["summary"] as? String ?? ""
        let memories = parsed["memories"] as? [String] ?? []
        
        var tasks: [TaskData] = []
        if let taskArray = parsed["tasks"] as? [[String: Any]] {
            for taskDict in taskArray {
                let taskTitle = taskDict["title"] as? String ?? ""
                let taskDescription = taskDict["description"] as? String
                var dueDate: Date?
                if let dueDateStr = taskDict["due_date"] as? String {
                    dueDate = dateFormatter.date(from: dueDateStr)
                }
                tasks.append(TaskData(title: taskTitle, description: taskDescription, dueDate: dueDate))
            }
        }
        
        return SummaryResult(title: title, summary: summary, memories: memories, tasks: tasks)
    }
}
